go/vumitools/middleware.py

```diff
@@ -603,6 +604,54 @@ def handle_failure(self, failure, connector_name):
         return failure
 
 
+class ConversationMetricsMiddlewareConfig(MetricsMiddlewareConfig):
+
+    redis_manager = ConfigDict(
+        "Redis configuration parameters", default={}, static=True)
+
+
+class ConversationMetricsMiddleware(MetricsMiddleware):
+    """
+    Middleware that stores which conversations have received or sent messages
+
+    :param dict redis_manager:
+        Connection configuration details for Redis.
+    """
+
+    CONFIG_CLASS = ConversationMetricsMiddlewareConfig
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        # We don't use a VumiApi here because we don't have a Riak config for
+        # it.
+        self.redis = yield TxRedisManager.from_config(
+            self.config.redis_manager)
+
+    def teardown_middleware(self):
+        return self.redis.close_manager()
+
+    @inlineCallbacks
+    def get_conv_key(self, msg):
+        mdh = MessageMetadataHelper(
+            self.vumi_api, msg, conversation_cache=self._conversation_cache)
+        conv_key = yield mdh.get_conversation_key()
+        returnValue(conv_key)
+
+    def record_conv_seen(self, msg):
+        conv_key = self.get_conv_key(msg)
```

Comment:
Yep, except that get_conv_key doesn't need to return a deferred, so it
could also not return one. Welcome to the condundrum of async programming
-- when to return a deferred. :)